File: csv_parser.cs
// csv_parser.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: dotnet run <input.csv> [output.json]");
            return;
        }

        string inputFile = args[0];
        string outputFile = args.Length > 1 ? args[1] : "output.json";

        Console.WriteLine("📊 CSV → JSON Парсер (C#)");

        if (!File.Exists(inputFile))
        {
            Console.WriteLine($"❌ Файл {inputFile} не найден.");
            return;
        }

        string content = File.ReadAllText(inputFile);
        var lines = content.Split('\n').Where(l => !string.IsNullOrWhiteSpace(l)).ToArray();
        if (lines.Length == 0)
        {
            Console.WriteLine("❌ CSV пуст.");
            return;
        }

        string delimiter = DetectDelimiter(lines[0]);
        Console.WriteLine($"🔍 Определён разделитель: {delimiter}");

        var headers = lines[0].Split(delimiter).Select(h => h.Trim()).ToArray();
        Console.WriteLine($"📋 Заголовки: {string.Join(", ", headers)}");

        var data = new List<Dictionary<string, object>>();
        for (int i = 1; i < lines.Length; i++)
        {
            var values = lines[i].Split(delimiter).Select(v => v.Trim()).ToArray();
            var record = new Dictionary<string, object>();
            for (int j = 0; j < headers.Length; j++)
            {
                string val = j < values.Length ? values[j] : "";
                record[headers[j]] = ParseValue(val);
            }
            data.Add(record);
        }

        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(outputFile, json);

        Console.WriteLine($"✅ Экспортировано {data.Count} записей в {outputFile}");
    }

    static string DetectDelimiter(string line)
    {
        int comma = line.Count(c => c == ',');
        int semicolon = line.Count(c => c == ';');
        int tab = line.Count(c => c == '\t');
        if (semicolon > comma && semicolon > tab) return ";";
        if (tab > comma && tab > semicolon) return "\t";
        return ",";
    }

    static object ParseValue(string value)
    {
        if (string.IsNullOrEmpty(value)) return null;
        string v = value.Trim().ToLower();
        if (v == "true") return true;
        if (v == "false") return false;
        if (v == "null" || v == "none") return null;
        if (int.TryParse(value, out int i)) return i;
        if (double.TryParse(value, out double d)) return d;
        return value;
    }
}

File: csv_parser.rs
// csv_parser.rs — Rust версия

use serde_json::json;
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader, Write};

fn detect_delimiter(line: &str) -> char {
    let comma = line.matches(',').count();
    let semicolon = line.matches(';').count();
    let tab = line.matches('\t').count();
    if semicolon > comma && semicolon > tab {
        ';'
    } else if tab > comma && tab > semicolon {
        '\t'
    } else {
        ','
    }
}

fn parse_value(value: &str) -> serde_json::Value {
    let v = value.trim();
    if v.is_empty() {
        return serde_json::Value::Null;
    }
    match v.to_lowercase().as_str() {
        "true" => return serde_json::Value::Bool(true),
        "false" => return serde_json::Value::Bool(false),
        "null" | "none" => return serde_json::Value::Null,
        _ => {}
    }
    if let Ok(i) = v.parse::<i64>() {
        return serde_json::Value::Number(i.into());
    }
    if let Ok(f) = v.parse::<f64>() {
        if let Some(n) = serde_json::Number::from_f64(f) {
            return serde_json::Value::Number(n);
        }
    }
    serde_json::Value::String(v.to_string())
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Usage: cargo run -- <input.csv> [output.json]");
        return Ok(());
    }

    let input_file = &args[1];
    let output_file = if args.len() > 2 {
        args[2].clone()
    } else {
        "output.json".to_string()
    };

    println!("📊 CSV → JSON Парсер (Rust)");

    let file = File::open(input_file)?;
    let reader = BufReader::new(file);
    let lines: Vec<String> = reader
        .lines()
        .filter_map(|line| {
            let l = line.ok()?;
            if l.trim().is_empty() { None } else { Some(l) }
        })
        .collect();

    if lines.is_empty() {
        println!("❌ CSV пуст.");
        return Ok(());
    }

    let delimiter = detect_delimiter(&lines[0]);
    println!("🔍 Определён разделитель: {}", delimiter);

    let headers: Vec<String> = lines[0]
        .split(delimiter)
        .map(|h| h.trim().to_string())
        .collect();
    println!("📋 Заголовки: {:?}", headers);

    let mut data = Vec::new();
    for line in &lines[1..] {
        let values: Vec<&str> = line.split(delimiter).map(|v| v.trim()).collect();
        let mut record = serde_json::Map::new();
        for (i, header) in headers.iter().enumerate() {
            let val = if i < values.len() { values[i] } else { "" };
            record.insert(header.clone(), parse_value(val));
        }
        data.push(serde_json::Value::Object(record));
    }

    let json = serde_json::to_string_pretty(&data)?;
    let mut file_out = File::create(&output_file)?;
    file_out.write_all(json.as_bytes())?;

    println!("✅ Экспортировано {} записей в {}", data.len(), output_file);
    Ok(())
}
